File: 205-IsomorphicStrings/205-IsomorphicStrings.java
// Last updated: 16/10/2025, 04:28:11
class Solution {
    public boolean isIsomorphic(String s, String t) {
        int s_len = s.length();
        int t_len = t.length();
        Map<Character, Character> mps = new HashMap<>();
        Map<Character, Character> mpt = new HashMap<>();

        if(s_len != t_len){
            return false;
        }

        int i = 0;
        while(i < s_len){
            if(mps.containsKey(s.charAt(i)) && mps.get(s.charAt(i)) != t.charAt(i)){
                return false;
            }
            else if(mpt.containsKey(t.charAt(i)) && mpt.get(t.charAt(i)) != s.charAt(i)){
                return false;
            }
            else{
                mps.put(s.charAt(i), t.charAt(i));
                mpt.put(t.charAt(i), s.charAt(i));
            }
            i++;
        }
        return true;
    }
}

File: 205-IsomorphicStrings/205-IsomorphicStrings.cpp
// Last updated: 16/10/2025, 04:28:17
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if(s.length() != t.length()) return false;
        int n = s.length();
        int m = t.length();
        unordered_map<char,char> mps;
        unordered_map<char,char> mpt;
        for(int i=0;i<n;i++){
            if(mps.find(s[i]) != mps.end()){
                if(mps[s[i]] != t[i]) return false;
            }
//we have to check from s->t as well as t->s. for ex s = "badc" && t= "baba".Try this test case.
            
            if(mpt.find(t[i]) != mpt.end()){
                if(mpt[t[i]] != s[i]) return false;
            }
            mps[s[i]] = t[i];
            mpt[t[i]] = s[i];
        }
        return true;
    }
};